### PR TITLE
Fix - 27030 - Validate Shipping rates

### LINKED ISSUE #27030: Products with no "Per Class" Flat Rates set for their class can be purchased if cart contains another class that does have rates
**Prerequisites (mark completed items with an \[x\]):**

- [X] I have have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**<br>When using "Per Class" calculations in the Flat Rate method (not "Per Order"), products of shipping classes that do not have rates entered are able to be purchased without a shipping cost, so long as other products in the cart have rates entered for their class.

A common use case affected by this is related to stores that have products that are only available if purchased from specific locations. In these situations, Flat Rates are only entered for the Shipping Classes that are available in each zone and unavailable products have the rate fields left blank. This approach currently works as expected when products are purchased individually, but when the cart contains both available and unavailable products, only the rates of the available product are used and purchase is able to be completed, rather than providing the "No shipping options were found" message.

**Expected behavior**<br>Since rates do not exist for all shipping classes, no rates should be offered.

**Actual behavior**<br>If at least one of the shipping classes in the cart have rate calculated, those costs are used and products of classes that do not have rates are ignored, allowing the purchase to be completed with incorrect shipping cost calculations.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**<br>Steps to reproduce the bug:

1. Create 2 Shipping Classes and 2 Shipping Zones (with Flat Rate added to each).
2. Use the "Per Class" calculation type and only set rates for 1 class in each zone. My example/screenshots use the "Austin" class with rates in the Austin Zone and "Texas" class with rates in the Texas Zone.
3. Calculate rates for the products individually, noticing rates are correct if they exist. If they do not exist in the matched zone, no method will be found for the class without rates entered.
4. Add both products to the cart and notice that only the rates that exist in the zone are used.

**Screenshots**<br>Example of Austin product being calculated for a zone with rates entered:

(Link to screenshot: [https://d.pr/i/ZinD8P+](<https://d.pr/i/ZinD8P+>))

Example of Texas product being calculated for a zone without rates entered:

(Link to screenshot: [https://d.pr/i/Y8iRSY+](<https://d.pr/i/Y8iRSY+>))

Example of both products being purchased at the same time and only the Austin class rates being used:

(Link to screenshot: [https://d.pr/i/eL0hJm+](<https://d.pr/i/eL0hJm+>))

**Isolating the problem (mark completed items with an \[x\]):**

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/storefront/>).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**<br>WordPress version 5.4.2<br>WooCommerce version 4.2.2<br>Storefront version 2.5.8

### PR DESCRIPTION
**Pull request by dpanta94**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #27030 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create 2 Shipping Classes and 2 Shipping Zones (with Flat Rate added to each, we could name them New York and Florida).
2. Use the "Per Class" calculation type and only set rates for 1 class in each zone.
3. Create 2 products one adding the New York shipping class and the Florida to the other one
4. Add them to the cart individually, you should be able to complete your checkout if your shipping address matches the product's shipping class.
5. Add them to the cart together, you should never be able to complete your checkout. You should see that there are no shipping options available.

### CODE REVIEW

**Comment by barryhughes** [comment]
Thanks for submitting the PR: as this touches a critical flow it's going to take a little while longer to finish review and testing, and it is probably important for us to add test coverage.

**Comment by dpanta94** [comment]
Hey @barryhughes ! Completely understandable, please let me know if you would like me to elaborate on anything regarding the logic i put in place. I tried to explain everything with comments, but i understand that some things may not be crystal clear.

Thank you for taking the time to review this!

**Comment by barryhughes** [comment]
@dpanta94 I appreciate your patience here! This is a really nice piece of work and we appreciate you taking the time to look into a fix. There are three things I'd like to discuss/flag before we take this further:

* Unit tests
* Unaddressed scenarios
* If we're solving this at the right level

The first is that we will absolutely need to add unit test coverage. As @juliaamosova pointed out [over here](https://github.com/woocommerce/woocommerce/issues/27030#issuecomment-657800206), this is a critical flow—so it's especially important we add tests both for safety and to describe how expect the system to behave. We can probably help with that if needed, just let us know (but, if so, it might need to wait on some bandwidth opening up).

The second is that there is at least one closely related scenario I think we should also address. Suppose that we have:

* Two shipping classes, A and B.
* A shipping zone with two flat rate methods.
* The first method charges $5 for class A (but ignores class B).
* The second method charges $10 for class B (but ignores class A).

Using our current code, if a shopper adds two products to the cart (one belonging to class A, the other to class B) then they will be able to choose between the two shipping methods. With this PR, they will only be presented with one of the shipping methods. It feels to me like neither behavior is correct, and we should again present a *"no shipping options available"* error. 

Last but definitely not least, I'm wondering if we're applying the fix at the right level. We already try to [calculate shipping for each package](https://github.com/woocommerce/woocommerce/blob/385d8e0fec9fc046932ef00d2a3a09c9421264ca/plugins/woocommerce/includes/class-wc-shipping.php#L260) and, of course, what this PR adds is an [extra validation step](https://github.com/woocommerce/woocommerce/blob/385d8e0fec9fc046932ef00d2a3a09c9421264ca/plugins/woocommerce/includes/class-wc-shipping.php#L275)—running immediately after the initial shipping calculation—which loops back over the package contents (generally, this means all the items in the cart) and builds ['virtual' packages](https://github.com/woocommerce/woocommerce/blob/385d8e0fec9fc046932ef00d2a3a09c9421264ca/plugins/woocommerce/includes/class-wc-shipping.php#L352) for each individual product, as a way to test if a shipping rate can be determined for that item. This is potentially quite a lot of extra work, that could be done more efficiently at an earlier stage.

Focusing on the flat rate shipping method for a moment, it seems that it may be possible to achieve the same outcome by adding an extra conditional inside [`WC_Shipping_Flat_Rate::calculate_shipping()`](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L179-L183) because, at this stage, we are able to determine all of the following:

- If the calculation type is *per class.*
- If a cost cannot be determined for an individual item in the package.

In other words: for those shipping methods that support working with shipping classes, might it be better to delegate this work to them? This means a different set of trade-offs, but I think it is worth discussing (and I know you've thought about this a great deal, so I'd love to hear your thoughts).

*...This developed into quite the wall of text, please let me know if I can clarify anything.*

**Comment by barryhughes** [comment]
...Expanding a little on one of those earlier thoughts: it seems the original bug report was specifically focused on the Flat Rate method. A potential drawback with the proposed strategy is that it would apply to *all* shipping methods, not just flat rate, and so the performance cost of repeated calls to `calculate_shipping_for_package()` could be problematic, particularly in the case of shipping methods that need to make a remote API call to do the calculation.

**Comment by dpanta94** [comment]
Hey @barryhughes,

Ill start addressing your points in the reverse order you mentioned them 😃

- Regarding if we're solving this at the right level.

I did consider solving this in a much earlier stage, but after investigating this deeper doing that in my opinion it would result in much more complex problems to solve. The earlier you apply changes, the hardest is to anticipate the results after all. This appeared as the safest option, considering the amount of extensions there are for WooCommerce. I did miss though considering a solution through [WC_Shipping_Flat_Rate::calculate_shipping()](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L179-L183) and to be honest this seems like a better approach. Please allow me some time, to work on this.

- Unaddressed scenarios

Hopefully i addressed most of them. At least, the example you mention i did take it under consideration and in my local works great. So with my PR, if we have a shipping zone with 2 flat rates. Each flat rate having applicable price for only one shipping class and we add 2 products in the cart each shipped via a shippping class matching a different flat rate of the shipping zone, the cart would display "No shipping options were found".

image
image

image
image
image
image
image

The only scenario i am aware its not addressed is cases where a shipping rate could be matched for more than one product. For example it could be cases where a shipping option becomes available after the cart's value is higher than 70$. As far as i am aware though these scenarios are not supported by WooCommmerce core, that's why the filter `woocommerce_skip_shipping_rates_validation` has been exposed.

- Regarding the extra load being added.

Yea an extra load is being added. I completely agree that it should be refactored so its not there for every possible shippping method and only for the flat rates. For the flat rates though an extra load seems necessary since the appropriate check resulting to this bug is missing. Considering there will be no external API calls for flat rates, the extra load should be minimum.

- Unit tests

I'd be happy to work on those. If you agree, i'd love if we finalized the approach as it seems some changes are necessary. I could need a little bit of guidance writting them. If i do i will reach out.


Let me know what you think! In the meantime, ill rework the approach applied in the WC_Shipping_Flat_Rate class.

**Comment by barryhughes** [comment]
Thanks, @dpanta94.

I'm not sure what the difference is because your setup for the scenario I described seems to closely match what I used—but in any case I think it may be something we can set aside, since:

> Considering a solution through [WC_Shipping_Flat_Rate::calculate_shipping()](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L179-L183) ... Please allow me some time, to work on this.

Definitely, no rush on our end. It does feel like this may be the better/simpler approach, unless we discover some reason it won't work.

Thanks again!

**Comment by barryhughes** [comment]
Just a friendly note that code freeze for the 7.8.0 release happens on Monday next week (2023-05-22). 

This means that, realistically, if we want to include this fix in the 7.8.0 release then the PR will need to be updated and ready for a fresh review early this week (of course, it can also wait until some future point—I just wanted to give you an idea of the timelines we are working to re 7.8.0).

**Comment by dpanta94** [comment]
Hey @barryhughes,

Would you mind taking a look on the refactored approach ? There is a test failing, but i dont see any connection to the altered code. Ill provide the tests as well.

Thank you!

File: plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php [hunk 1]
```diff
@@ -173,11 +173,41 @@ public function calculate_shipping( $package = array() ) {
...
+
+			// Store this is the first iteration of shipping classes loop.
+			$first_iteration = true;
+
 			foreach ( $found_shipping_classes as $shipping_class => $products ) {
 				// Also handles BW compatibility when slugs were used instead of ids.
 				$shipping_class_term = get_term_by( 'slug', $shipping_class, 'product_shipping_class' );
 				$class_cost_string   = $shipping_class_term && $shipping_class_term->term_id ? $this->get_option( 'class_cost_' . $shipping_class_term->term_id, $this->get_option( 'class_cost_' . $shipping_class, '' ) ) : $this->get_option( 'no_class_cost', '' );
 
+				foreach ( $products as $data ) {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php]: I'm wondering if we can avoid this inner loop 🤔 

Looking at the [current implementation](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L146-L217), we only add a rate if we've determined that [there are shipping costs](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L206-L208) and, for that [to be true](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L185), we must have discovered a shipping class that yields a [non-empty class cost string](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L181).

If the class cost string is empty, **and** if the calculation type is per class (ie, `'class' === $this->type`), perhaps that is all the reason we need to just bail out of the `$found_shipping_classes` loop? Something like this:

```php
$no_class_cost = 0 === strlen( $class_cost_string );

if ( $no_class_cost && 'class' === $this->type ) {
	// If we are calculating per each individual shipping class, but one of the products belongs to a
	// class for which a cost has not been defined, then we cannot supply a shipping rate.
	$has_costs = false;
	break;
} elseif ( $no_class_cost ) {
	// Otherwise, if we do not have a class cost but we are intending to use the highest available
	// rate, we should simply skip this product.
	continue;
}
```

If this is viable, it would be more efficient (since we don't need to loop through all of the products).

Comment (dpanta94) [reply plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php]: Yes @barryhughes you were right. Missed this. Checked it out and it works great and its definitely more performant.

**Comment by barryhughes** [comment]
Thanks for the updates: I have some high priority items I am attending to first, but will try to circle back to this as soon as I can.

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php [hunk 2]
```diff
@@ -611,6 +611,190 @@ public function test_calculate_shipping() {
 		$this->assertEquals( 20, $object->get_shipping_total() );
 	}
 
+	public function test_calculate_shipping_with_shipping_classes() {
```
Comment (barryhughes) [inline plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: Thanks for adding this! However, we should probably move it to a new location. As [noted here](https://github.com/woocommerce/woocommerce/blob/7.7.2/plugins/woocommerce/tests/README.md#writing-tests):

> - `tests/legacy/unit-tests` contains tests for code in the includes directory. No new tests should be added here, ever; existing test classes shouldn't get new tests either. Fixing faulty existing tests is allowed.
> - `tests/php/includes` is where all the new tests for code in the includes directory should be written.

In this case we are testing code in the includes directory, so the test belongs in `tests/php/includes`. I suggest a new path:

`tests/php/includes/shipping/flat-rate/class-wc-shipping-flat-rate-test.php`

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php [hunk 3]
```diff
@@ -611,6 +611,190 @@ public function test_calculate_shipping() {
...
+		$instance_id_3 = $zone_3->add_shipping_method( 'flat_rate' );
+		$instance_id_4 = $zone_3->add_shipping_method( 'flat_rate' );
+		$zone_3->save();
+
+		$map = array(
+			$shipping_class_1['term_id'] => $instance_id_1, // NY
+			$shipping_class_2['term_id'] => $instance_id_2, // FL
+			$shipping_class_3['term_id'] => $instance_id_3, // CA
+			$shipping_class_4['term_id'] => $instance_id_4, // IL
+		);
```
Comment (barryhughes) [inline plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: (1/3) I'm a little concerned about overall readability once we reach the first assertion. What if we instead define the map in this way:

```php
$method_1_cost = '125';
/* Other method costs... */

$map = array(
	$shipping_class_1['term_id'] => array( $instance_id_1, $method_1_cost ),
	/* ... */
);

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php [hunk 4]
```diff
@@ -611,6 +611,190 @@ public function test_calculate_shipping() {
...
+				'woocommerce_flat_rate_cost'                                       => '',
+				'woocommerce_flat_rate_class_cost_' . $shipping_class_1['term_id'] => '',
+				'woocommerce_flat_rate_class_cost_' . $shipping_class_2['term_id'] => '',
+				'woocommerce_flat_rate_class_cost_' . $shipping_class_3['term_id'] => '',
+				'woocommerce_flat_rate_class_cost_' . $shipping_class_4['term_id'] => '',
+				'woocommerce_flat_rate_type'                                       => 'class',
+				'instance_id'                                                      => $instance_id,
+			);
+
+			$data['woocommerce_flat_rate_class_cost_' . $shipping_class_id ] = $shipping_class_id * 10;
```
Comment (barryhughes) [inline plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: (2/3) Then, here, we could change this loop to something like this:

```php
foreach ( $map as $shipping_class_id => $method ) {
	list( $method_id, $method_cost ) = $method;

	$data = array(
		/* As currently defined... */
		'instance_id' => $method_id,
	);

	$data['woocommerce_flat_rate_class_cost_' . $shipping_class_id ] = $method_cost;
	/* Rest of loop... */
}
```

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php [hunk 5]
```diff
@@ -611,6 +611,190 @@ public function test_calculate_shipping() {
...
+		WC_Helper_Shipping::force_customer_us_address();
+
+		WC()->cart->empty_cart();
+
+		WC()->cart->add_to_cart( $product_1->get_id(), 1 );
+
+		WC()->cart->calculate_shipping();
+
+		// Customers address in in NY and product_1 is shippable to NY.
+		$this->assertEquals( (string) ( 10 * $shipping_class_1['term_id'] ), WC()->cart->get_shipping_total() );
```
Comment (barryhughes) [inline plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: (3/3) Finally, and building on the earlier 2 comments in the series, this lets us avoid the arithmetic at this point:

```php
$this->assertEquals( $method_1_cost, WC()->cart->get_shipping_total() );
```

Additionally and optionally, for this and other assertions we could move the comment into the description string:

```php
$this->assertEquals( 
	$method_1_cost, 
	WC()->cart->get_shipping_total(),
	'Customers address in in NY and product_1 is shippable to NY.'
);

**Review by barryhughes — CHANGES_REQUESTED** [review]
Sorry for the delay here, but this is looking great! I do have some hopefully final feedback regarding the test you added (the most important thing is that we move it to a new location), but otherwise I think we're very close to merge.

**Comment by barryhughes** [comment]
Please also note the code sniff results, we need those to pass in order to merge (though it makes sense to relocate the test first of all).

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php (hunk 2 above)
Comment (dpanta94) [reply plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: Sounds good @barryhughes! Created the file and transferred the code over there!

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php (hunk 3 above)
Comment (dpanta94) [reply plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: Yea makes sense, done!

File: plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php (hunk 4 above)
Comment (dpanta94) [reply plugins/woocommerce/tests/legacy/unit-tests/order/class-wc-tests-crud-orders.php]: Done as well!

**Comment by dpanta94** [comment]
@barryhughes amended according to your suggestions. Also applied fixes for the phpcs action to succeed. The action Run unit tests on PR / PHP 8.0 WP latest is failing, but it seems unrelated to the proposed changes.

**Comment by barryhughes** [comment]
Thanks for the updates! I triggered a re-run of the E2E test, though I think you are right: it seems unrelated.

Edit: the PHP unit test though might be flagging something valid, just looking at that one.

**Comment by barryhughes** [comment]
This is a tricky one:

```
Run unit tests on PR / PHP 8.0 WP latest (pull_request) Failing after 12m

WC_Shipping_Flat_Rate_Test::test_calculate_shipping_with_shipping_classes
Undefined array key "line_total"
```

I can replicate it locally if I run the entire suite (PHP 8.0 and WP 6.2.2) but *not* if I run that same test in isolation. Possibly the result of leaky tests elsewhere, but it still makes sense for us to solve it—otherwise, we'll cause problems for others after merging.

It may take a little while to figure it out, however. Bear with us!

**Comment by barryhughes** [comment]
Skipping `WC_Tests_Core_Functions::test_wc_load_cart()` solves the problem when the test suite runs under PHP 8.0, still a little unsure why that would be.

**Comment by dpanta94** [comment]
@barryhughes i see that you have identified the issue and fixed and the PR has been merged in #39146. Thats awesome! As a result, ill be closing this one! Thank you very much!

**Comment by barryhughes** [comment]
No problem, thank you for working on this one - it's a really nice fix.
